Generalized Store System

EmoGarbage404 opened the pull request:
<!-- The text between the arrows are comments - they will not be visible on your PR. -->
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->
this is going to migrate uplinks to this new generic system. This will also allow for more robust conditions for uplink listings as well as using this same system for new antags/systems which need to use an uplink-like shop.

Closes #8855 
Closes #9559 
Closes #2772

**Screenshots**
<!-- If applicable, add screenshots to showcase your PR. If your PR is a visual change, add
screenshots or it's liable to be closed by maintainers. -->

**Changelog**
<!--
Here you can fill out a changelog that will automatically be added to the game when your PR is merged
There are 4 icons for changelog entries: add, remove, tweak, fix. I trust you can figure out the rest.

You can put your name after the :cl: symbol to change the name that shows in the changelog (otherwise it takes your GitHub username)
Like so: :cl: PJB

Generally, only put things in changelogs that players actually care about. Stuff like "Refactored X system, no changes should be visible" shouldn't be on a changelog.

For writing actual entries, don't consider the entry type suffix (e.g. add) to be "part" of the sentence:
bad: - add: a new tool for engineers
good: - add: added a new tool for engineers
-->

:cl:
- add: The Syndicate Zombie Bundle has been readded for Nuclear Operatives for 40 TC.
- tweak: Nuclear Operatives can no longer purchase Holoparasites or Surplus bundles.
- tweak: Empty Uplink categories no longer show up in the menu.

Just-a-Unity-Dev commented:
>no more holoparasites

oh boy, cant wait to see the outrage in lizard OOC

EmoGarbage404 commented:
> > no more holoparasites
> 
> oh boy, cant wait to see the outrage in lizard OOC

jokes on you i never play so i will never see their rage.
regardless, the bundles are the bigger nerf tbh.

asperger-sind commented:
> no more surpluses

Finally, the day where nukies actually need to choose their loadout and ask each-other for TC has come, gone are the days of 5/5 nukies just buying super surpluses and having enough equipment to arm a small country and 4 balloons, gone are the days of shuttle looting (well, partially ig?)

moonheart08 commented on Content.Server/Store/Components/CurrencyComponent.cs:
```diff
@@ -0,0 +1,13 @@
+namespace Content.Server.Store.Components;
+
+[RegisterComponent]
+public sealed class CurrencyComponent : Component
```
This component and all fields need explanatory doc comments.

moonheart08 commented on Content.Server/Store/Components/StoreComponent.cs:
```diff
@@ -0,0 +1,76 @@
+namespace Content.Server.Store.Components;
+
+[RegisterComponent]
+public sealed class StoreComponent : Component
```
This component needs a doc comment.

moonheart08 commented on Content.Server/Store/Components/StoreComponent.cs:
```diff
@@ -0,0 +1,76 @@
+    [ViewVariables]
+    public HashSet<ListingData> LastAvailableListings = new();
+
+    #region audio
```
All fields in this region need doc comments.

moonheart08 commented on Content.Server/Store/Components/StoreComponent.cs:
```diff
@@ -0,0 +1,76 @@
+    #endregion
+}
+
+public sealed class StoreAddedEvent : EntityEventArgs { };
+public sealed class StoreRemovedEvent : EntityEventArgs { };
```
Doc comments.

moonheart08 commented on Content.Server/Store/Conditions/StoreAntagCondition.cs:
```diff
@@ -0,0 +1,61 @@
+/// Supports both blacklists and whitelists. This is copypaste because roles
+/// are absolute shitcode. Refactor this later. -emo
+/// </summary>
+public sealed class StoreAntagCondition : ListingCondition
```
Fields and functions of this need doc comments. Remember `<inheritdoc/>` exists for overrides.

moonheart08 commented on Content.Server/Store/Conditions/StoreJobCondition.cs:
```diff
@@ -0,0 +1,61 @@
+/// Allows a store entry to be filtered out based on the user's job.
+/// Supports both blacklists and whitelists
+/// </summary>
+public sealed class StoreJobCondition : ListingCondition
```
Fields and functions of this need doc comments. Remember `<inheritdoc/>` exists for overrides.

moonheart08 commented on Content.Server/Store/Conditions/StoreLimitedStockCondition.cs:
```diff
@@ -0,0 +1,17 @@
+/// <summary>
+/// Only allows a listing to be purchased a certain amount of times.
+/// </summary>
+public sealed class StoreLimitedStockCondition : ListingCondition
```
Fields and functions of this need doc comments. Remember `<inheritdoc/>` exists for overrides.

moonheart08 commented on Content.Server/Store/Conditions/StoreUserWhitelistCondition.cs:
```diff
@@ -0,0 +1,35 @@
+/// <summary>
+/// Filters out an entry based on the components or tags on an entity.
+/// </summary>
+public sealed class StoreUserWhitelistCondition : ListingCondition
```
Fields and functions of this need doc comments. Remember `<inheritdoc/>` exists for overrides.

moonheart08 commented on Content.Server/Store/Systems/StoreSystem.Listings.cs:
```diff
@@ -0,0 +1,86 @@
+
+namespace Content.Server.Store.Systems;
+
+public sealed partial class StoreSystem : EntitySystem
```
This system and all functions need thorough doc comments.

moonheart08 commented on Content.Server/Store/Systems/StoreSystem.Ui.cs:
```diff
@@ -0,0 +1,201 @@
+
+namespace Content.Server.Store.Systems;
+
+public sealed partial class StoreSystem : EntitySystem
```
This system and all functions need thorough doc comments.

moonheart08 commented on Content.Server/Store/Systems/StoreSystem.cs:
```diff
@@ -0,0 +1,123 @@
+
+namespace Content.Server.Store.Systems;
+
+public sealed partial class StoreSystem : EntitySystem
```
This system and all functions need thorough doc comments.

moonheart08 commented on Content.Server/Traitor/Uplink/UplinkSystem.cs:
```diff
@@ -234,11 +31,20 @@ public bool AddUplink(EntityUid user, UplinkAccount account, EntityUid? uplinkEn
+
+            var currency = new Dictionary<string, FixedPoint2>
+            {
+                { "Telecrystal", balance }
```
Isn't there a TcCurrencyPrototype defined somewhere?

EmoGarbage404 replied:
"Telecrystal" is the ID of that prototype. UplinkSystem doesn't have a component so I suppose i could just make it a const.

moonheart08 commented on Content.Shared/Store/CurrencyPrototype.cs:
```diff
@@ -0,0 +1,28 @@
+/// </summary>
+[Prototype("currency")]
+[DataDefinition, Serializable, NetSerializable]
+public sealed class CurrencyPrototype : IPrototype
```
Fields need a doc. Not the name and ID though those are fine due to being **genuinely** self-explanatory.

moonheart08 commented on Content.Shared/Store/ListingCondition.cs:
```diff
@@ -0,0 +1,13 @@
+
+[ImplicitDataDefinitionForInheritors]
+[MeansImplicitUse]
+public abstract class ListingCondition
```
Doc comments.

moonheart08 commented on Content.Shared/Store/ListingPrototype.cs:
```diff
@@ -0,0 +1,102 @@
+/// </summary>
+[Serializable, NetSerializable]
+[Virtual, DataDefinition]
+public class ListingData : IEquatable<ListingData>
```
Document your fields!!!!

moonheart08 commented on Content.Shared/Store/StoreCategoryPrototype.cs:
```diff
@@ -0,0 +1,22 @@
+    public string Name { get; } = string.Empty;
+
+    [DataField("priority")]
+    public int Priority { get; } = 10;
```
Why 10? Shouldn't this just be 0?

moonheart08 commented on Content.Shared/Store/StorePresetPrototype.cs:
```diff
@@ -0,0 +1,35 @@
+/// </summary>
+[Prototype("storePreset")]
+[DataDefinition]
+public sealed class StorePresetPrototype : IPrototype
```
Doc comments for fields!

EmoGarbage404 commented:
conflicts resolved and moony's review addressed.

PJB3005 commented on Resources/Locale/en-US/store/currency.ftl:
```diff
@@ -0,0 +1,9 @@
+store-currency-inserted = {CAPITALIZE(THE($used))} is inserted into the {THE($target)}.
+
+store-currency-free = Free
+store-currency-name-debugdollar = {$amount ->
+    [one] Debug Dollar
+    *[other] Debug Dollars
+}
+
+store-currency-name-telecrystal = TC
```
This doesn't make much sense to me: one is an acronym, another is a full name. Shouldn't the bottom be "Telecrystals"?

EmoGarbage404 replied:
I suppose these are a bit of a misnomer. They aren't the display name of the currency, but rather the text used for displaying the price in the ui itself (i.e.: listing has a button that says [10 TC]). In that sense, it makes sense to use the shorter TC abbreviation for display, just like how the old uplink system has it.

mirrorcult commented on Content.Server/Store/Components/StoreComponent.cs:
```diff
@@ -0,0 +1,92 @@
+    [ViewVariables(VVAccess.ReadOnly), DataField("currencyWhitelist", customTypeSerializer: typeof(PrototypeIdHashSetSerializer<CurrencyPrototype>))]
+    public HashSet<string> CurrencyWhitelist = new();
+
+    /// <summary>
+    /// Whether or not this store can be activated by clicking on it (like an uplink)
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite), DataField("activateInHand")]
+    public bool ActivateInHand = true;
```
instead of having this why not just specify `ActivatableUI` in YAML?

EmoGarbage404 replied:
This needs to be here because the stores have to use their unique logic for getting all of the listings and such when opened. I'm not sure if that can be done through the generic ActivatableUI component. This is also how the original uplinks did it as well.

EmoGarbage404 replied:
wait i'm actually dumb lemme just implement this

mirrorcult commented on Content.Server/Store/Conditions/StoreAntagCondition.cs:
```diff
@@ -0,0 +1,63 @@
+/// Supports both blacklists and whitelists. This is copypaste because roles
+/// are absolute shitcode. Refactor this later. -emo
+/// </summary>
+public sealed class StoreAntagCondition : ListingCondition
```
stuff like supersurplus being attached to `StoreAntagCondition` IMO doesn't make a lot of sense. this means that if a non-nukie gets ahold of a syndicate radio, they can use it to order a super surplus crate/holopara because they don't actually have the antag role. i really feel like it should be just similar to `StoreUserWhitelist` but for the uplink entity, and then the radio can have a special tag that blacklists nukie items

mirrorcult reviewed (changes requested):
rename folders `Ui` -> `UI`

are you going to be updating the revenant store to use this in this PR or a diff PR?

in testing `adduplink` no longer works, even if I specify the PDA ID? it also seems like the syndicate radios no longer work, at least the ones i tried had `activateInHand` being false

EmoGarbage404 commented:
fixed the broken command, renamed the folder, and fixed the issues with the radios not activating.  I'm going to update revenant store in a different pr (as this one is quite large already)

EmoGarbage404 commented:
the namespace is staying as `Content.Client.Store.Ui` because github and my computer have organized a shadow cabal to ruin my life.
